### ISSUE TITLE
Feature request: allow ignoring characters

### ISSUE THREAD

**Issue by jyn514** [maciejhirsz/logos#106]
The C language requires that backslashes followed by a newline should be removed - not just treated as whitespace, but ignored entirely as if they were never there. For example, the following code is valid:

```
i +\
= 1;
```
but this is not: `i + = 1;` (notice the space between `+` and `=`).

I have a custom lexer in `rcc` and this is the only major thing stopping me from switching to logos. The way I currently handle this in rcc is by consuming the newline _before_ parsing any tokens: https://github.com/jyn514/rcc/blob/master/src/lex/mod.rs#L91. However, doing this with logos would require making an entirely separate pass to remove the `\\\n` before ever passing the input to the logos lexer. Is it possible to add this feature to logos?

If I'm missing an existing way to do this, please do let me know.

**Comment by Evrey** [comment]
Addendum: It's not just `\LF`, it's `\\\n[[:c_whitespace:]]*`, for whatever C considers as being valid whitespace. I.e. this feature needs to be able to handle regexes. As an attribute name I propose `line_noise`, as this is pretty much what this boils down to. =)

**Comment by maciejhirsz** [comment]
can this occur in the middle of things like identifiers, as in: is `foo\_bar` expected to produce `foobar` where `_` is a newline?

**Comment by jyn514** [comment]
Yes, that's right.

**Comment by jyn514** [comment]
> Addendum: It's not just `\LF`, it's `\\\n[[:c_whitespace:]]*`, for whatever C considers as being valid whitespace.

This is not correct, it only deletes the newline and backslash.

**Comment by Evrey** [comment]
Huh, okay then. Guess I confused this with how Rust does multi line strings.

**Comment by maciejhirsz** [comment]
In that case doing a first pass is probably the best thing to do. You could do a simple `&str` -> `Cow<str>` conversion that runs through the source checking for `b"\\\n"` (super easy to do with `.as_bytes().windows(2).find(...`, or just `.find("\\\n")`). Happy path here (no `b"\\\n"` found) could just re-borrow the original `&str` in its entirety without having to copy anything. I don't know what is the statistical % of C files that use this, but there has to be a non-zero number of C files/programs that never use it :).

Adding this specifically to logos I can see two problems:
1. Logos makes pretty strong assumptions about being able to read and match source in uninterrupted chunks up to 16 bytes. There might be some way to add support for alternative [`Source`](https://docs.rs/logos/0.10.0/logos/source/trait.Source.html) that only reads one byte at a time with a filter, but that would render quite a few optimizations Logos does useless, and it's plausible that net performance would be slower than doing a first pass. In fact, if first pass is doing a happy path I think that's guaranteed.
2. For things like identifiers you probably want to be able to borrow them from the source in a form that is actually useful in AST or whatever other thing you are parsing into. If you do do 1. and skip over problematic bytes, then you probably then have to sanitize all your source slices to filter those bytes out (again). This can be an external cost to the lexer, but it's still a cost somewhere down in the parser.

You could try to implement the first pass filter, and remove the filter from your current lexer, and see how that affects performance. If it doesn't degrade it, or doesn't degrade it _too much_ for sources that do use `"\\\n"`, then I reckon it might be worth it to make the shift. If/when Logos starts to exploit SIMD it might be worth it even first pass is degrading perf.

One obvious drawback to having a first pass preparser is that your byte offsets wouldn't match to original source, so there would have to be some translation layer.

**Comment by maciejhirsz** [comment]
Come to think of it, you co do first pass that doesn't have to do copying and doesn't break byte offsets. If you have a mutable buffer (`String` or `&mut str` or some such), you could, in principle, do something like this (assuming `_` is newline):

```
"i +\_= 1;"
"i +=_  1;"
    ^^^
```

Highlighted are 3 bytes that get affected. The algorithm is roughly as follows: whenever you find a sequence of `"\\\n"`, shift all non-whitespace bytes that follow it into it's position, and put `"\n "` in place you are shifting from.

**Comment by jyn514** [comment]
> Happy path here (no b"\\\n" found) could just re-borrow the original &str in its entirety without having to copy anything. I don't know what is the statistical % of C files that use this, but there has to be a non-zero number of C files/programs that never use it :).

I expect most programs will use it, so it's almost always going to copy. The use case here is for multiline macros: preprocessor macros must be on only one line, so any program that wants them to be split into multiple has to use this trick. For example, llvm (~7 million LOC) uses this 70000 times:

```
$ tokei
 Language            Files        Lines         Code     Comments       Blanks
-------------------------------------------------------------------------------
 C                    6017       876374       483325       301806        91243
 C Header             7731      1483430       922710       340659       220061
 C++                 20702      5381635      3828063       883105       670467
 C++ Header             31        11581         9655          863         1063
$ rg '\\$' | wc -l
68085
```

> One obvious drawback to having a first pass preparser is that your byte offsets wouldn't match to original source, so there would have to be some translation layer.

This is unfortunate but I could probably work around it. I'd only need the original byte offsets for pretty-printing, so it's ok if whatever algorithm I used was slow.

> The algorithm is roughly as follows: whenever you find a sequence of "\\\n", shift all non-whitespace bytes that follow it into it's position, and put "\n " in place you are shifting from.

Hmm, interesting. If I replaced it with `"\n\n"` it would keep the same byte offset as well. The issue here is what happens if there is no whitespace following, like `int main() { i+\_=1;}`. Also, this looks like it would require arbitrary lookaheads into the file, and as mentioned it would break strings.

 So I think overall it would be better to do this in a separate pass and keep a BTreeSet<usize> or similar of where the line breaks were, and then pass the output to logos.

**Comment by Evrey** [comment]
> I don't know what is the statistical % of C files that use this

I don't know whether that's CPP or C grammar rules, but `\LF` is extremely common in `#define`s.

> Logos makes pretty strong assumptions about being able to read and match source in uninterrupted chunks up to 16 bytes.

That'll be an issue. And one cannot easily `Span` a filtering iterator, as all indices will be off. That mutable buffer technique is nifty, though.

**Comment by maciejhirsz** [comment]
> So I think overall it would be better to do this in a separate pass and keep a `BTreeSet<usize>` or similar of where the line breaks were, and then pass the output to logos.

I reckon you can get away with `Vec<usize>` and use `binary_search` from std.

**Comment by windelbouwman** [comment]
Interesting! I had exactly this question (in fact I started with an experiment called `rcc` myself, working on preprocessing mostly).

I faced two missing features in logos:

1. Handling the above mentioned backslash + newline glue action as described in the C standard. This actually means that this can occur anywhere.
2. Support for locations other than `usize`, but more a `row, column` pair. This would greatly help if this was available from logos right away. This is not part of this issue, so I will not further discuss here.

Valid C snippets with the backslash newline:

```C
/\
* Some \
comments on 1 line */

```

Another one:

```C
int fu\
bar = 2;

printf("Fubar=%d\n", fubar);
```

In fact, the C translation phases are described here:

https://en.cppreference.com/w/c/language/translation_phases

The logos lexer would be used in phase 3.

One idea that might work here is the following: Use a pre filtering of the sourcecode in which you search for newlines and backslash + newline combinations.
- Whenever you find a backslash + newline combo, increase a `new_lines_counter`.
- Whenever a newline is found, inject `new_lines_counter` extra newlines, and reset `new_lines_counter` to zero.

This messes up row / column information for tokens after the backslash + newline, but other tokens have correct line information.
